frame/staking/src/lib.rs

```diff
@@ -3335,19 +3335,44 @@ impl<T: Config> Module<T> {
 	}
 }
 
+use sp_election_providers::data_provider::ReturnValue;
+
 impl<T: Config> sp_election_providers::ElectionDataProvider<T::AccountId, T::BlockNumber>
 	for Module<T>
 {
-	fn desired_targets() -> u32 {
-		Self::validator_count()
+	type Additional = Weight;
+	fn desired_targets() -> ReturnValue<u32, Self::Additional> {
+		Ok((Self::validator_count(), <T as frame_system::Config>::DbWeight::get().reads(1)))
 	}
 
-	fn voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		Self::get_npos_voters()
+	fn voters(
+		maybe_max_len: Option<usize>,
+	) -> ReturnValue<Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)>, Self::Additional> {
+		let nominator_count = <Nominators<T>>::iter().count();
+		let validator_count = <Validators<T>>::iter().count();
+		let voter_count = nominator_count.saturating_add(validator_count);
+
+		if maybe_max_len.map_or(false, |max_len| voter_count > max_len) {
+			return Err("Voter snapshot too big");
+		}
+
+		// NOTE: in the grand scheme of things, given that we iterate all validators and only
+		// 'alive' validators can have slashing spans, we ought to read all slashing spans at least
+		// once, given overlay cache EXACTLY once from the database.
+		let read_count = voter_count.saturating_add(<SlashingSpans<T>>::iter().count());
+		let weight = <T as frame_system::Config>::DbWeight::get().reads(read_count as u64);
+		Ok((Self::get_npos_voters(), weight))
```

Comment:
Actually I'm not sure doing next_key is cached.
We put into cache what keys contains what value.
But not what is the next key of one key.
I think we should factorize the counting in get_npos_voters.

Reply:
Can you elaborate more? I don't think I make any assumption about caching the next key. The read count is the number of all voters (which is the db read count of `<Nominators>::iter()` without any caching), plus all the slashing spans once. 

Effectively, I think what we do is simply equal to `|<Nominators>::iter()| ++ |<SlashingSpans>::iter()|`

Reply:
ahh I see, I forgetting `<Validators>` here, is this what you mean?

Reply:
ah no the code looked correct I was thinking about this line:
```
let nominator_count = <Nominators<T>>::iter().count();
```
Doing this require use to iterate on all key in the storage `Nominators` which I thought can be quite costy.
So doing it 2 times may be not wanted.

Currently we don't count the number of `next_key` calls. But every call `next_key` go into the trie and find the next key. Maybe we should be careful about it.

Reply:
Hmm interesting, then I can actually refactor this to not use count. I can just build manual counters in and return them manually. 

But the missing validators is indeed wrong. we also iterate those in the `get_npos_voters `. I can make an assumption that since we know they are being called at the same block, we won't count them in.

Reply:
hmm also I am forgetting about all the reads to `Ledger` and `Bonded`, this is wrong and needs proper benchmarking.

Reply:
> Currently we don't count the number of next_key calls. But every call next_key go into the trie and find the next key. Maybe we should be careful about it.

I wrote a dummy benchmark and doing `.count` indeed is `O(N)`, and doing `.count` followed by `iter().collect()` of the same thing is still `O(N)`

Reply:
what kind of benchmark? I thought benchmark define in `benchmark!` macro doesn't really count for the cost of trie operations, that's why we count storage read and writes independently, no ?